pkg/fleets/controller.go

```diff
@@ -309,19 +314,119 @@ func (c *Controller) deleteEmptyGameServerSets(fleet *stablev1alpha1.Fleet, list
+	replicas, err := c.rollingUpdateActive(fleet, active, rest)
+	if err != nil {
+		return replicas, err
+	}
+	if err := c.rollingUpdateRest(fleet, rest); err != nil {
+		return replicas, err
+	}
+	return replicas, nil
+}
+
+// rollingUpdateActive applies the rolling update to the active GameServerSet
+// and returns what its replica value should be set to
+func (c *Controller) rollingUpdateActive(fleet *stablev1alpha1.Fleet, active *stablev1alpha1.GameServerSet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	replicas := active.Spec.Replicas
+
+	// if the active spec replicas are greater than or equal the fleet spec replicas, then we don't
+	// need to another rolling update upwards.
+	// Likewise if the active spec replicas don't equal the active status replicas, this means we are
+	// in the middle of a rolling update, and should wait for it to complete.
+	if active.Spec.Replicas >= fleet.Spec.Replicas || active.Spec.Replicas != active.Status.Replicas {
+		return replicas, nil
+	}
+
+	r, err := intstr.GetValueFromIntOrPercent(fleet.Spec.Strategy.RollingUpdate.MaxSurge, int(fleet.Spec.Replicas), true)
+	if err != nil {
+		return replicas, errors.Wrapf(err, "error calculating scaling gameserverset: %s", fleet.ObjectMeta.Name)
+	}
+	surge := int32(r)
+
+	// make sure we don't end up with more than the configured max surge
+	maxSurge := surge + fleet.Spec.Replicas
+	replicas = fleet.UpperBoundReplicas(replicas + surge)
+	total := stablev1alpha1.SumStatusReplicas(rest) + replicas
+	if total > maxSurge {
+		replicas = fleet.LowerBoundReplicas(replicas - (total - maxSurge))
+	}
+
+	// always leave room for Allocated GameServers
+	sumAllocated := stablev1alpha1.SumStatusAllocatedReplicas(rest)
+
+	// make room for allocated game servers, but not over the fleet replica count
+	if replicas+sumAllocated > fleet.Spec.Replicas {
+		replicas = fleet.LowerBoundReplicas(replicas - sumAllocated)
+	}
+
+	c.logger.WithField("gameserverset", active.ObjectMeta.Name).WithField("replicas", replicas).
+		Info("applying rolling update to active gameserverset")
+
+	return replicas, nil
+}
+
+// rollingUpdateRest applies the rolling update to the inactive GameServerSets
+func (c *Controller) rollingUpdateRest(fleet *stablev1alpha1.Fleet, rest []*stablev1alpha1.GameServerSet) error {
+	if len(rest) == 0 {
+		return nil
+	}
+
+	r, err := intstr.GetValueFromIntOrPercent(fleet.Spec.Strategy.RollingUpdate.MaxUnavailable, int(fleet.Spec.Replicas), true)
+	if err != nil {
+		return errors.Wrapf(err, "error calculating scaling gameserverset: %s", fleet.ObjectMeta.Name)
+	}
+	unavailable := int32(r)
+
+	for _, gsSet := range rest {
+		// if the status.Replicas are less than or equal to 0, then that means we are done
+		// scaling this GameServerSet down, and can therefore exit/move to the next one.
+		if gsSet.Status.Replicas <= 0 {
```

Comment:
@enocom the inversion here breaks this out into two sections, with slightly different logic - which I think is much easier to read :+1: and parse. WDYT?